packages/core/src/controller/iterator-controller.ts

```diff
@@ -17,7 +22,30 @@ export class IteratorController<TOut> implements Controller<TOut>, Trapper {
     this.resume(() => this.iterator.next());
   }
 
-  resume(iter: () => IteratorResult<Operation<unknown>>) {
+  // the purpose of this method is solely to make `step` reentrant, that is we
+  // should be able to handle a `halt` which occurs while we are already in a
+  // generator. This is a rare case and should only happen under some edge
+  // cases, for example a task halting itself, or a task causing one of its
+  // siblings to fail.
+  resume(iter: Continuation) {
+    if(this.didEnter) {
+      // if we're already in a `step`, store the continuation for later instead
+      // of running it immediately.
+      this.continuations.push(iter);
+    } else {
+      this.didEnter = true; // acquire lock
+      this.step(iter);
+      this.didEnter = false; // release lock
+
+      // resume with stored continuation it happened while we were in the step
+      let continuation = this.continuations.shift();
+      if(continuation) {
+        this.resume(continuation)
+      }
+    }
+  }
```

Comment:
This definitely makes the iterator controller a lot less "elegant", but it is the simplest fix that I could come up with.

Reply:
@jnicklas I wonder if we can convert this to a loop that is just draining the continuations until there are none?

Reply:
Yeah, making this recursive is maybe a bit strange. I did not want to use `forEach` to prevent problems with the collection being modified during iteration, but a while loop would totally work as well.

Reply:
Ahh, I think I see what you're saying, basically you want something like this:

``` typescript
  resume(iter: Continuation) {
    this.continuations.push(iter);
    let continuation;
    while(continuation = this.continuations.shift()) {
      this.step(continuation);
    }
  }
```

Unfortunately this doesn't work, since we need to somehow ensure only one of these loops is running at a time.

Basically we need to build a really crappy mutex here 😆

Reply:
This is an alternate implementation, I'm not sure if this is better, what do you think?

``` typescript
  resume(iter: Continuation) {
    this.continuations.push(iter);
    if(!this.didEnter) {
      this.didEnter = true; // acquire lock
      let continuation;
      while(continuation = this.continuations.shift()) {
        this.step(continuation);
      }
      this.didEnter = false; // release lock
    }
  }
```

Reply:
I like that implementation better, let's go with that!